storage/index.go

```diff
@@ -265,16 +264,57 @@ func (i *nsIndex) Query(
 		return index.QueryResults{}, err
 	}
 
-	idxIter := index.NewIterator(i.nsID, iter, i.opts)
+	var (
+		brokeEarly = false
+		numAdded   = 0
+		results    = i.opts.ResultsPool().Get()
+		cleanupFn  = func() {
+			results.Finalize()
+			exec.Close()
+			iter.Close()
+		}
+	)
+	results.Reset(i.nsID)
+
+	for iter.Next() {
+		// FOLLOWUP(prateek): as part of the runtime options wiring, also set a server side max limit which
+		// super-seeds the user requested limit. Paranoia can be a good thing.
```

Comment:
If we have pagination though it may be ok to do large scrolls for instance.

Reply:
You'd want different server side limits for fetch v scroll but fair.